### PR TITLE
Add SatImp method

### PR DESCRIPTION
**Pull request by Puning97**
We add the SatImp (ICML2025) method for open-unlearning framework. 5 files are changed.

### CODE REVIEW

**Comment by molereddy** [comment]
Can you link to the papers you are implementing?

**Comment by Puning97** [comment]
> Can you link to the papers you are implementing?

Of course! In fact, I’ve already added the relevant links in the README.md and in docs/links.

| SatImp               | Paper[📄](https://arxiv.org/pdf/2505.11953), Code [🐙](https://github.com/Puning97/SatImp-for-LLM-Unlearning)                                                                                      |
| WGA (G-effect)       | Paper[📄](https://arxiv.org/pdf/2502.19301), Code [🐙](https://github.com/tmlr-group/G-effect)                                                                                                     |

**Comment by molereddy** [comment]
Sorry I meant in the PR description. It's hard to navigate to these.

**Comment by Puning97** [comment]
> Sorry I meant in the PR description. It's hard to navigate to these.

Sorry for the misunderstanding. I guess you mean that I need to modify the top PR description, adding these links so that others can easily find them? Am I correct?

Now the PR is:
We add the SatImp (ICML2025) method for open-unlearning framework. 5 files are changed.

I need to modify to:

## What does this PR do?
This PR introduces the integration of the SatImp and WGA methods.
The original papers are:

SatImp: [arxiv](https://arxiv.org/pdf/2505.11953),   [Github](https://github.com/Puning97/SatImp-for-LLM-Unlearning)
WGA: [arxiv](https://arxiv.org/pdf/2502.19301), [Github](https://github.com/tmlr-group/G-effect) 

Have you gone through the contributions [guide](https://github.com/locuslab/open-unlearning/docs/contributing.md)?  Yes

Are your changes documented? Read documentation guidelines [here](https://github.com/locuslab/open-unlearning/README.md#-further-documentation). Yes

**Comment by molereddy** [comment]
It seems the ruff check tests we have (for code formatting) are failing. Please follow the instructions in [`docs/contributing.md`](https://github.com/locuslab/open-unlearning/blob/main/docs/contributing.md#create-a-pull-request) and re-format the code. 
Also, just to re-confirm, since we will not be testing before merging, 
1. this exact version of the codebase 
2. in an enviroment with the package versions described in the readme
3. is working without any errors

Thank you a lot for contributing!

**Comment by Puning97** [comment]
> Aside from code formatting, LGTM.

Sorry for the multiple attempts! The reason is the difference between ruff==0.12.0 and ruff==0.6.6, which makes the different format in utils.py!

File: src/trainer/unlearn/satimp.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+            shift_logits.view(-1, shift_logits.size(-1)), shift_labels.view(-1)
+        )
+        weight_sat = ((-lm_loss).exp().detach()) ** self.beta1
+        weight_imp = (1 - (-lm_loss).exp().detach()) ** self.beta2
+        forget_loss = -((weight_sat * weight_imp) * lm_loss)[
+            shift_labels.view(-1) != -100
+        ].mean()
+        return forget_loss
+
+    def compute_retain_loss(self, model, retain_inputs):
```
Comment (Dornavineeth) [inline src/trainer/unlearn/satimp.py:37]: Would it be better if you could inherit from GradDiff and skip re-writing the 
compute_retain_loss function?

File: src/trainer/unlearn/satimp.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    ):  # attention, satimp requires two beta!!!!
+        super().__init__(*args, **kwargs)
+        self.beta1 = beta1
+        self.beta2 = beta2
+        self.gamma = gamma
+        self.alpha = alpha
+        if self.ref_model is None:
+            self.ref_model = self._prepare_ref_model(self.model)
+
+    def compute_satimp_loss(self, model, forget_inputs):
```
Comment (Dornavineeth) [inline src/trainer/unlearn/satimp.py:18]: To enable your loss function usage with other complementary approaches like optimizer-based unlearning works, you can implement a function in [utils.py](https://github.com/locuslab/open-unlearning/blob/main/src/trainer/utils.py). See [compute_undial_loss](https://github.com/locuslab/open-unlearning/blob/f05ad6cc05d10381eb10d8bbc124f051b6475fd3/src/trainer/utils.py#L71).

File: src/trainer/unlearn/wga.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        shift_logits = outputs.logits[..., :-1, :].contiguous()
+        shift_labels = labels[..., 1:].contiguous()
+        lm_loss = nn.CrossEntropyLoss(ignore_index=-100, reduction="none")(
+            shift_logits.view(-1, shift_logits.size(-1)), shift_labels.view(-1)
+        )
+        weight_ce = ((-lm_loss).exp().detach()) ** self.beta
+        forget_loss = -(weight_ce * lm_loss)[shift_labels.view(-1) != -100].mean()
+        return forget_loss
+
+    def compute_retain_loss(self, model, retain_inputs):
```
Comment (Dornavineeth) [inline src/trainer/unlearn/wga.py:31]: Same comment as https://github.com/locuslab/open-unlearning/pull/125/files#r2173634028

File: src/trainer/unlearn/wga.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+class WGA(UnlearnTrainer):
+    def __init__(self, beta=1.0, gamma=1.0, alpha=1.0, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.gamma = gamma
+        self.alpha = alpha
+        self.beta = beta
+        if self.ref_model is None:
+            self.ref_model = self._prepare_ref_model(self.model)
+
+    def compute_wga_loss(self, model, forget_inputs):
```
Comment (Dornavineeth) [inline src/trainer/unlearn/wga.py:15]: Same comment as https://github.com/locuslab/open-unlearning/pull/125/files#r2173635168

File: configs/trainer/SatImp.yaml [hunk 5]
```diff
@@ -0,0 +1,15 @@
+defaults:
+  - finetune
```
Comment (Dornavineeth) [inline configs/trainer/SatImp.yaml:2]: If the change https://github.com/locuslab/open-unlearning/pull/125/files#r2173634028 is done, don't forget to import from graddiff here instead of finetune.

File: configs/trainer/WGA.yaml [hunk 6]
```diff
@@ -0,0 +1,14 @@
+defaults:
+  - finetune
```
Comment (Dornavineeth) [inline configs/trainer/WGA.yaml:2]: If the change https://github.com/locuslab/open-unlearning/pull/125/files#r2173634028 is done, don't forget to import from graddiff here instead of finetune.

Comment (Puning97) [reply configs/trainer/WGA.yaml]: Thanks for you suggestion! I have updated all of files! Please check them again!

**Comment by Puning97** [comment]
> @Puning97 Congratulations on your two new method additions to Open-Unlearning 🎉🎉🎉! We look forward to seeing more of your advanced work and contributions to OU.

Thank you for your valuable contributions too. I look forward to continued collaboration on advanced research in the future.
